File: lib/data/api/jamendo_api.dart
import 'package:dio/dio.dart';

class JamendoApi {
  final Dio _dio;
  final String clientId;

  JamendoApi(this.clientId) : _dio = Dio(BaseOptions(
    baseUrl: 'https://api.jamendo.com/v3.0',
    connectTimeout: const Duration(seconds: 15),
    receiveTimeout: const Duration(seconds: 15),
  ));

  Future<Map<String, dynamic>> searchTracks(String query, {int limit = 20}) async {
    final response = await _dio.get('/tracks', queryParameters: {
      'client_id': clientId,
      'format': 'json',
      'limit': limit,
      'search': query,
    });
    return response.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> getTracks({int limit = 20}) async {
    final response = await _dio.get('/tracks', queryParameters: {
      'client_id': clientId,
      'format': 'json',
      'limit': limit,
    });
    return response.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> getTracksByGenre(String genre, {int limit = 20}) async {
    final response = await _dio.get('/tracks', queryParameters: {
      'client_id': clientId,
      'format': 'json',
      'limit': limit,
      'tags': genre,
    });
    return response.data as Map<String, dynamic>;
  }
}

File: app/src/main/java/com/soundflow/app/data/remote/api/JamendoApi.kt
package com.soundflow.app.data.remote.api

import com.soundflow.app.data.remote.dto.JamendoArtistsResponse
import com.soundflow.app.data.remote.dto.JamendoTracksResponse
import retrofit2.http.GET
import retrofit2.http.Query

interface JamendoApi {
    @GET("tracks/")
    suspend fun getTracks(
        @Query("client_id") clientId: String,
        @Query("format") format: String = "json",
        @Query("limit") limit: Int = 20,
        @Query("offset") offset: Int = 0,
        @Query("order") order: String = "popularity_total",
        @Query("include") include: String = "musicinfo",
        @Query("audioformat") audioFormat: String = "mp32"
    ): JamendoTracksResponse

    @GET("tracks/")
    suspend fun searchTracks(
        @Query("client_id") clientId: String,
        @Query("format") format: String = "json",
        @Query("limit") limit: Int = 20,
        @Query("namesearch") query: String,
        @Query("include") include: String = "musicinfo",
        @Query("audioformat") audioFormat: String = "mp32"
    ): JamendoTracksResponse

    @GET("tracks/")
    suspend fun getTracksByGenre(
        @Query("client_id") clientId: String,
        @Query("format") format: String = "json",
        @Query("limit") limit: Int = 20,
        @Query("tags") genre: String,
        @Query("order") order: String = "popularity_total",
        @Query("include") include: String = "musicinfo",
        @Query("audioformat") audioFormat: String = "mp32"
    ): JamendoTracksResponse

    @GET("artists/")
    suspend fun searchArtists(
        @Query("client_id") clientId: String,
        @Query("format") format: String = "json",
        @Query("limit") limit: Int = 20,
        @Query("namesearch") query: String
    ): JamendoArtistsResponse
}
